Keck HIRES continues

profxj opened the pull request:
This is another substantial PR on the path to fully enabling
Keck/HIRES reductions.

The code is now running end-to-end on the one test dataset
in the DevSuite.

I am submitting now to:

- Get eyes on the modest changes to the `WaveCalib` datamodel
- Have the refactoring of `ech_objfind()` ingested

This PR should not merge until after PR #1527

profxj committed:
towards object finding

Commit:
still fussing

Commit:
working without manual

Commit:
cleaning

Commit:
hand..

Commit:
multiple wv2dfit

Commit:
writing to disk

Commit:
Merge remote-tracking branch 'origin/develop' into hires_continues

Commit:
broke something

Commit:
Merge remote-tracking branch 'origin/fire_fixes' into hires_continues

Commit:
fussin

Commit:
Merge remote-tracking branch 'origin/fire_fixes' into hires_continues

Commit:
runnin

Commit:
mo

Commit:
doc me

Commit:
way mo docs

Commit:
done with find docs

Commit:
mo polishing

Commit:
wavecalib

Commit:
angles

Commit:
Merge remote-tracking branch 'origin/develop' into hires_continues

Commit:
changes

profxj commented:
DevSuite tests are underway

jhennawi commented on pypeit/find_objects.py:
```diff
@@ -1010,26 +1009,103 @@ def find_objects_pypeline(self, image, ivar, std_trace=None,
-            nabove_min_snr=self.par['reduce']['findobj']['ech_find_nabove_min_snr'],
             box_radius=self.par['reduce']['extraction']['boxcar_radius'],  # arcsec
-            show_trace=show_trace, objfindQA_filename=objfindQA_filename)
+            objfindQA_filename=objfindQA_filename)
```
Echelle object finding is a core functionality just like non-echelle object finding. The only object dependency is on specobjs. By adding the control flow that was in core here, one now needs all objects needed to instantiate find_objects, which adds more dependencies. Also, someone may just want to find objects but not sky-subtract, which this class does not allow one to do in a simple way. I don't see a current motivation for taking this echelle object finding control flow outside of core.

jhennawi commented on pypeit/wavecalib.py:
```diff
@@ -39,28 +38,34 @@ class WaveCalib(datamodel.DataContainer):
+                 'wv_fit2d': dict(otype=np.ndarray, atype=fitting.PypeItFit,
+                                  descr='2D wavelength solution(s) (echelle).  If there is more than one, they must be aligned to the separate detectors analyzed'),
+                 'det_img': dict(otype=np.ndarray, atype=np.integer,
+                                  descr='Detector image; used occasionally by echelle.  Currently only saved if ech_separate_2d=True'),
```
Can you add a few more words indicating what the detector image is. I.e. is it an image with an integer indicating the detector that contributed the pixel, etc.

profxj replied:
done

profxj replied:
done

jhennawi reviewed (changes requested):
Nice work on this @profxj. The results look great. The only significant comment is the refactoring of the echelle object finding. I think that control flow should remain a function in core and not be moved into the find_objects class. 

Also, since the code was completely refactored, I cannot see what changes if any were actually made to the echelle object finding functionality via a normal side-by-side code review.  It may be useful to chat about what was changed if anything significant was changed.

Commit:
mo

Commit:
WIP

Commit:
ready to try

Commit:
debugging

Commit:
doc

Commit:
Merge remote-tracking branch 'origin/develop' into hires_continues

Commit:
working!

Commit:
Merge remote-tracking branch 'origin/develop' into hires_continues

Commit:
polishing

kbwestfall commented on pypeit/core/wavecal/echelle.py:
```diff
@@ -181,8 +185,9 @@ def identify_ech_orders(arcspec, echangle, xdangle, dispname, angle_fits_file, c
-    pad : int
+    pad : int, optional
         Number of orders to pad the coverage by on the blue and red side.
+    debug : bool, optional
```
Give a short description.

profxj replied:
added

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -982,7 +982,7 @@ def initial_trace(self, bpm=None):
 
         # Only after the correction for the bad column, update the bmp with the
-        # regions to exclude from slit tracing
+        # - regions to exclude from slit tracing
```
?

profxj replied:
linting was otherwise going bonkers.  no idea why..

kbwestfall commented on pypeit/images/buildimage.py:
```diff
@@ -26,7 +26,9 @@ class ArcImage(pypeitimage.PypeItImage):
     # I/O
-    output_to_disk = ('ARC_IMAGE', 'ARC_FULLMASK', 'ARC_DETECTOR')
+    output_to_disk = ('ARC_IMAGE', 'ARC_FULLMASK', 'ARC_DETECTOR',
+                      'ARC_DET_IMG', # For echelle multi-detector wavelengths
```
Do we need to save this image?  Or can we only require that we're able to reproduce it using the saved "DETECTOR" object?  If it's the image I'm thinking of, this should be fast to construct.

profxj replied:
I considered that, but to enable `WaveCalib`
to generate the wavelength image it needs
this and it may not have the rest of what is needed
to regenerate.

kbwestfall commented on pypeit/par/pypeitpar.py:
```diff
@@ -2469,6 +2470,16 @@ def __init__(self, reference=None, method=None, echelle=None, ech_fix_format=Non
+        descr['ech_fix_format'] = 'Is this a fixed format echelle?  If so reidentification ' \
+                                  'will assume that each order in the data is aligned with a ' \
+                                  'single order in the reid arxiv.'
+
```
I know you only moved this code, but I'm taking the opportunity to say that I don't understand why this is a parameter and not an attribute of the relevant spectrograph.  I.e., why do we need to toggle whether or not a spectrograph is a fixed-format echelle?

_Edit_: Actually, it looks like we do define this at least for the `KeckHIRESSpectrgraph`.  Can we consolidate?

profxj replied:
indeed, that `par` was superfluous
now removed

kbwestfall commented on pypeit/slittrace.py:
```diff
@@ -1104,6 +1105,43 @@ def assign_maskinfo(self, sobjs, plate_scale, spat_flexure, TOLER=1.):
+        # Remove 0s and find most common
+        det_vals = det_vals[det_vals > 0].tolist()
+        data = Counter(det_vals)
+        ordr_det = data.most_common(1)[0][0]
```
I'm not sure which is faster, but it seems like you could also use `numpy.unique`, setting `return_counts=True`, and return the unique value with the most counts.  The code would then be:
```
det, cnt = np.unique(det_img[(slit_img == spat_id) & (det_img > 0)], return_counts=True)
return det[np.argmax(cnt)]
```

profxj replied:
love it!  replaced

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -1011,26 +1010,103 @@ def find_objects_pypeline(self, image, ivar, std_trace=None,
-            nabove_min_snr=self.par['reduce']['findobj']['ech_find_nabove_min_snr'],
-            box_radius=self.par['reduce']['extraction']['boxcar_radius'],  # arcsec
-            show_trace=show_trace, objfindQA_filename=objfindQA_filename)
+        orig = True
```
Please add a big bright flashing boldface underlined TODO that this needs to be settled before we merge to `release`.

profxj replied:
removing `orig` now

jhennawi commented on pypeit/find_objects.py:
```diff
@@ -1011,26 +1010,103 @@ def find_objects_pypeline(self, image, ivar, std_trace=None,
-            nabove_min_snr=self.par['reduce']['findobj']['ech_find_nabove_min_snr'],
-            box_radius=self.par['reduce']['extraction']['boxcar_radius'],  # arcsec
-            show_trace=show_trace, objfindQA_filename=objfindQA_filename)
+        orig = True
```
Are you planning to delete the orig=False control flow below?

profxj replied:
yes

jhennawi commented on pypeit/spectrographs/keck_hires.py:
```diff
@@ -59,7 +59,7 @@ class KECKHIRESSpectrograph(spectrograph.Spectrograph):
     pypeline = 'Echelle'
     ech_fixed_format = False
-    supported = True
+    supported = False
```
Should this be True now?

profxj replied:
I don't think yet, e.g. we haven't implemented anything
sensible for flat fielding.  Am going to add notes here
on what else needs to be done before making it supported

jhennawi commented on pypeit/spectrographs/keck_hires.py:
```diff
@@ -515,38 +516,84 @@ def get_detector_par(self, det, hdu=None):
             numamplifiers   = 1,
-            # TODO THese are place holders at present. Getting these numbers right requires parsing headers to determine
-            # which readout mode is being used.
             gain            = np.atleast_1d([1.]),
```
is this correct, i.e. gain is exactly 1.0? Also, I believe this depends on readout mode so we probably should put a TODO in there about that.

profxj replied:
set below

jhennawi commented on pypeit/spectrographs/keck_hires.py:
```diff
@@ -515,38 +516,84 @@ def get_detector_par(self, det, hdu=None):
+            ronoise=np.atleast_1d([3.1])
         ))
 
+        # Set gain 
```
Now I'm confused. Rather than instantiate to bogus values and then set them again below, should we instantiate to None?

profxj replied:
sure

jhennawi commented on pypeit/wavecalib.py:
```diff
@@ -548,19 +574,19 @@ def build_wv_calib(self, arccen, method, skip_QA=False):
                                              nsnippet=self.par['nsnippet'])
                                              #debug=True, debug_reid=True, debug_xcorr=True)
         elif self.par['method'] == 'echelle':
```
Can you add a TODO somewhere about merging this eventually with "reidentify" for fixed format echelles?

profxj replied:
added

jhennawi reviewed (approved):
Hi @profxj, 

Thanks for the work here. I have a minor comment about the gain. The more significant comment is regarding the ech_objfind core routine. I like the changes you made there, and approve provided you don't revert back to the orig=False where that control flow is in the find_objects object.

profxj commented:
--- PYTEST PYPEIT UNIT TESTS PASSED  206 passed, 137 warnings in 191.42s (0:03:11) ---
--- PYTEST UNIT TESTS PASSED  118 passed, 242 warnings in 885.83s (0:14:45) ---
--- PYTEST VET TESTS PASSED  30 passed, 10 warnings in 762.34s (0:12:42) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 1/167 TESTS  ---
Failed tests:
    keck_hires/RED_C1_ECH_-0.82_XD_1.62 pypeit
Skipped tests:
Testing Started at 2023-02-17T05:15:50.969380
Testing Completed at 2023-02-17T13:50:11.598081
Total Time: 8:34:20.628701

Failed the main test! :)

kbwestfall commented on pypeit/wavecalib.py:
```diff
@@ -241,28 +257,38 @@ def build_waveimg(self, tilts, slits, spat_flexure=None, spec_flexure=None):
+                    ordr_det = slits.det_of_slit(
+                        slit_spat, self.det_img,
+                        slit_img=slit_img)
+                    idx_fit2d = ordr_det-1  # There are ways for this to go sour..
```
Can you articulate how in a comment so that we can look out for failure modes?

profxj replied:
added

kbwestfall commented on pypeit/wavecalib.py:
```diff
@@ -611,8 +636,8 @@ def build_wv_calib(self, arccen, method, skip_QA=False):
 
-    # TODO: Point to the datamodel for wv_calib in the docstring
-    def echelle_2dfit(self, wv_calib, debug=False, skip_QA=False):
+    # TODO -- Remove this
```
If you don't want to remove it now, can you comment it out for now?

profxj replied:
removed

kbwestfall commented on pypeit/wavecalib.py:
```diff
@@ -676,6 +699,108 @@ def echelle_2dfit(self, wv_calib, debug=False, skip_QA=False):
+        msgs.info('Fitting 2-d wavelength solution for echelle....')
+
+        # Obtain a list of good slits
+        ok_mask_idx = np.where(np.invert(self.wvc_bpm))[0]
```
(As annoying as it is) Can you use `logical_not` instead of `invert`?

profxj replied:
done

kbwestfall commented on pypeit/wavecalib.py:
```diff
@@ -676,6 +699,108 @@ def echelle_2dfit(self, wv_calib, debug=False, skip_QA=False):
+
+        # Prep
+        if self.par['ech_separate_2d']:
+            slit_img = self.slits.slit_img()
```
Is there a way to avoid having to recreate this so many times, but still not have to save it to the slit trace file?

profxj replied:
I don't think so.  And because it needs to allow for flexure, etc.
this is a better approach.

kbwestfall commented on pypeit/wavecalib.py:
```diff
@@ -676,6 +699,108 @@ def echelle_2dfit(self, wv_calib, debug=False, skip_QA=False):
+            # Init
+            all_wave = np.array([], dtype=float)
+            all_pixel = np.array([], dtype=float)
+            all_order = np.array([],dtype=float)
```
Will the elements of each of these empty arrays (i.e., the things you append below) always be the same shape?  If not, you might instead set `dtype=object`.  The allows the object to be an array of arrays.  I'm not sure how this is treated otherwise.

profxj replied:
good point, although as it ain't broke I'm not going to fix it.  ;)

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -147,7 +147,1159 @@ def create_skymask(sobjs, thismask, slit_left, slit_righ, box_rad_pix=None, trim
+        # Modify to 1-based indexing
+        obj_id_init = inobj_id + 1
+    elif nfound==1:
+        obj_id_init = np.ones(1,dtype='int')
```
Can `nfound` be 0?  If so, add an else here that throws an error?

profxj replied:
Good call

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -147,7 +147,1159 @@ def create_skymask(sobjs, thismask, slit_left, slit_righ, box_rad_pix=None, trim
+        slit_spat_id (`numpy.ndarray`_):
+            slit_spat values (spatial position 1/2 way up the detector)
+            for the orders 
+        std_trace (:class:`~pypeit.specobjs.SpecObjs`, optional): _description_. Defaults to None.
```
Does `_description_` do something?

profxj replied:
removed

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -147,7 +147,1159 @@ def create_skymask(sobjs, thismask, slit_left, slit_righ, box_rad_pix=None, trim
+    SNR_arr_trim = SNR_arr[:,keep_obj]
+
+    sobjs_final = sobjs_trim.copy()
+    return sobjs_final
```
Skip the middle man: `return sobjs_trim.copy()`?

profxj replied:
agreed

profxj replied:
and no `.copy()` necessary

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -147,7 +147,1159 @@ def create_skymask(sobjs, thismask, slit_left, slit_righ, box_rad_pix=None, trim
+        for iord, spec in enumerate(sobjs_final[indx_obj_id]):
+            # JFH added the condition on ech_frac_was_fit with S/N cut on 7-7-19.
+            # TODO is this robust against half the order being masked?
+            if spec.ech_frac_was_fit & (spec.ech_snr > 1.0):
```
Don't use `&` with single booleans.  Instead: 

`if spec.ech_frac_was_fit and spec.ech_snr > 1:`

profxj replied:
modified

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -147,7 +147,1159 @@ def create_skymask(sobjs, thismask, slit_left, slit_righ, box_rad_pix=None, trim
+        hand_extract_dict=manual_extract_dict)
+
+    # Lots of additional work for slits with sources (found or input manually)
+    if len(sobjs_in_orders) > 0: # or manual_extract_dict is not None:
```
Instead of this `if..else` block, I would immediately return the `else` case so that the rest of it doesn't need to be within an if statement.  I.e.:
```
if len(sobjs_in_orders) == 0:
    return sobjs_in_orders

...

return ech_pca_traces(...)
```

profxj replied:
good call

kbwestfall reviewed (changes requested):
Nice!  Nothing major from me.  I like the breaking up of the original function (at least what I was able to follow).  I put in a few comments that I realized afterward are likely hold-overs from the previous code.

Commit:
Merge remote-tracking branch 'origin/develop' into hires_continues

Commit:
PR edits

profxj commented:
Restarting tests

profxj commented:
Passed all the tests except HIRES!  :) 

(pypeit) wolverine> tail xavier-keck-hires-3.report 
--- PYTEST PYPEIT UNIT TESTS PASSED  206 passed, 137 warnings in 211.40s (0:03:31) ---
--- PYTEST UNIT TESTS PASSED  118 passed, 242 warnings in 985.68s (0:16:25) ---
--- PYTEST VET TESTS PASSED  30 passed, 9 warnings in 865.38s (0:14:25) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 1/167 TESTS  ---
Failed tests:
    keck_hires/RED_C1_ECH_-0.82_XD_1.62 pypeit
Skipped tests:
Testing Started at 2023-02-19T07:53:14.048351
Testing Completed at 2023-02-19T16:35:53.630866
Total Time: 8:42:39.582515

I introduced a bug updating `gain` handling as per the PR recommendation.
Fixed now

Commit:
gain fix

profxj commented:
Passed:

(pypeit) wolverine> ./pypeit_test -i keck_hires -t 4 reduce
Running reduce tests.
Running tests on the following instruments:
    keck_hires

Reducing data from keck_hires for the following setups:
    RED_C1_ECH_-0.82_XD_1.62

Running tests in 4 parallel processes
 0 passed/ 0 failed/ 0 skipped STARTED keck_hires/RED_C1_ECH_-0.82_XD_1.62 pypeit
 1 passed/ 0 failed/ 0 skipped PASSED  keck_hires/RED_C1_ECH_-0.82_XD_1.62 pypeit
--- PYPEIT DEVELOPMENT SUITE PASSED 1/1 TESTS  ---

Test Summary
--------------------------------------------------------
--- PYPEIT DEVELOPMENT SUITE PASSED 1/1 TESTS  ---
Testing Started at 2023-02-19T13:18:51.559156
Testing Completed at 2023-02-19T14:50:15.635608
Total Time: 1:31:24.076452
